napari/components/_layer_slicer.py

```diff
@@ -0,0 +1,138 @@
+from __future__ import annotations
+
+import logging
+from concurrent.futures import Executor, Future, ThreadPoolExecutor
+from threading import RLock
+from typing import Dict, Iterable, Optional, Tuple
+
+from napari.components import Dims
+from napari.layers import Layer
+from napari.utils.events.event import EmitterGroup, Event
+
+logger = logging.getLogger("napari.components._layer_slicer")
+
+
+class _LayerSlicer:
+    """
+    High level class to control the creation of a slice (via a slice request),
+    submit it (synchronously or asynchronously) to a thread pool, and emit the
+    results when complete.
+
+    Events
+    ------
+    ready
+        emitted after slicing is done with a dict value that maps from layer to slice response
+    """
+
+    def __init__(self):
+        """
+        Attributes
+        ----------
+        _executor : concurrent.futures.ThreadPoolExecutor
+            manager for the slicing threading
+        _layers_to_task : dict
+            task storage for cancellation logic
+        _lock_layers_to_task : threading.RLock
+            lock to guard against changes to `_layers_to_task` when finding,
+            adding, or removing tasks.
+        """
+        self.events = EmitterGroup(source=self, ready=Event)
+        self._executor: Executor = ThreadPoolExecutor(max_workers=1)
+        self._layers_to_task: Dict[Tuple[Layer], Future] = {}
+        self._lock_layers_to_task = RLock()
```

Comment:
why this is not used when sending task?

Reply:
It should be. @kcpevey : can you make the update to acquire the lock when adding the task?